gordon/router.py

```diff
@@ -0,0 +1,158 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Core message routing logic for the plugins within Gordon Service.
+
+Messages received on the success channel will be routed to the next
+designated plugin phase. For example, a message that has a ``consume``
+phase will be routed to the installed enricher provider (or publisher
+provider if no enricher provider is installed).
+
+If a message fails its next phase, its phase will be updated to ``drop``
+and routed to the event consumer provider for cleanup.
+
+.. attention::
+
+    The :class:`GordonRouter` only supports the following two phase
+    routes:
+
+    1. consume -> enrich -> publish -> done
+    2. consume -> publish -> done
+
+    Future releases may support more configurable phase routings.
+
+"""
+
+import asyncio
+import logging
+
+
+# TODO/NOTE (lynn): This is a temporary helper function to alias
+#                   entrypoints of plugins to a common name. In the
+#                   future, plugin interfaces will be updated to have a
+#                   common name for their entrypoint.
+def _set_plugin_entrypoint_aliases(plugin_dict):
+    # e.g. enricher_inst.process => enricher_inst.handle_message
+    entrypoints = {
+        'enricher': 'process',
+        'publisher': 'publish_changes',
+        'event_consumer': 'cleanup',
+    }
+    for plugin_name, func in entrypoints.items():
+        try:
+            plugin = plugin_dict[plugin_name]
+        except KeyError:  # if no enricher
+            continue
+        entry_point = getattr(plugin, func)
+        setattr(plugin, 'handle_message', entry_point)
+    return plugin_dict
+
+
+class GordonRouter:
+    """Route messages to the appropriate plugin destination.
+
+    .. attention::
+
+        `error_channel` is currently not used in this router, and may
+        be removed entirely from all interface definitions.
+
+    Args:
+        success_channel (asyncio.Queue): A sink for successfully
+            processed :class:`gordon.interfaces.IEventMessage` s.
+        error_channel (asyncio.Queue): A sink for
+            :class:`gordon.interfaces.IEventMessage` s that were not
+            processed due to problems.
+        plugins (dict): Implemented plugin provider names mapped to
+            their instantiated objects.
+    """
+    # TODO (lynn): Ideally this is configurable/extendable in future
+    #              iterations of gordon.
+    PHASE_TO_PLUGIN_MAPPER = {
+        'enrich': 'enricher',
+        'publish': 'publisher',
+        'done': 'event_consumer',
```

Comment:
How about `cleanup`? since that's the method name attached to `event_consumer` that will be called?

Reply:
I feel like that might be too closely tied, since messages in 'drop' will also go to cleanup right?  I'm assuming the reason 'drop' and 'done' are separate here is so its obvious for implementers that they *can* do different things.  If one is named the same as the method they both go to, it might be more confusing why 'drop' is there as a separate phase.